### PR TITLE
Improve and move gRPC documentation

### PR DESCRIPTION
**Pull request by Bargsteen**
## Purpose

Closes https://github.com/concordium/concordium-node/issues/369 (which has a more elaborate description)
Also closes https://github.com/Concordium/concordium-node/issues/264

## Changes

- Add gRPC reference page
- Update links on the SDK/api page
- Add JSON schemas generated from this branch: https://github.com/Concordium/concordium-rust-sdk/tree/derive-schema
- Explain how to generate these schemas in the readme

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by dg-concordium — APPROVED** [review]
I made one minor grammatical correction in grpc.rst. Otherwise it is fine. Nice use of the collapse function.

**Comment by dg-concordium** [comment]
In your intro I can see you have updated the SDKs. Can you also please update them here? https://developer.concordium.software/en/mainnet/net/guides/sdks-apis.html This is the landing page for the Develop on Concordium card on the front page. And the grpc one should obviously point to your grpc.rst topic.

File: source/mainnet/net/references/grpc.rst [hunk 1]
```diff
@@ -0,0 +1,886 @@
...
+
+Notation
+========
+
+This page uses the following notation:
+
+- ``?a`` means ``a`` OR ``null``.
+
+  - JSON schemas for ``?a``-types, will only specify the ``a``. But the node may
+    also return ``null``.
```
Comment (abizjak) [inline source/mainnet/net/references/grpc.rst:40]: I don't understand this. Won't it say the field is optional?

File: source/mainnet/net/references/grpc.rst [hunk 2]
```diff
@@ -0,0 +1,886 @@
...
+
+- ``[a]`` means a list of type ``a``.
+- JSON Schemas follow the `JSON Schema (Draft 7) <https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01>`_.
+- Protobuf code follow the `Protocol Buffers Version 3 Language Specification <https://developers.google.com/protocol-buffers/docs/reference/proto3-spec>`_.
+- ``Foo ::= (n: UInt16) (bars: Byteⁿ)``
+
+  - Denotes the serialization of the type ``Foo``, which starts with an unsigned
+    16-bit Integer, ``n``, which describes the number of bytes used to represent
```
Comment (abizjak) [inline source/mainnet/net/references/grpc.rst:48]: In big endian.

File: source/mainnet/net/references/grpc.rst [hunk 3]
```diff
@@ -0,0 +1,886 @@
...
+
+Transactions
+============
+
+.. function:: SendTransaction(networkId, payload) -> Bool
+
+   Send a transaction to the given network. See more information about the
+   payload in the :ref:`grpc-transaction-encoding` section.
+
+   :param Int32 networkId: The network that the transaction should be sent to (default is 100).
```
Comment (abizjak) [inline source/mainnet/net/references/grpc.rst:59]: I recommend stating "only 100 is currently supported".

File: source/mainnet/net/references/grpc.rst [hunk 4]
```diff
@@ -0,0 +1,886 @@
...
+   :param payload: Binary encoding of the transaction payload.
+   :type payload: |grpc-block-item|_
+   :returns: Whether the transaction succeeded.
+   :rtype: Bool
```
Comment (abizjak) [inline source/mainnet/net/references/grpc.rst:63]: I recommend stating in the description that the node will do basic transaction validation such as signature checks, nonce checks, and if these fail, the call will return a gRPC error.

File: source/mainnet/net/guides/developer-page.rst [hunk 5]
```diff
@@ -53,3 +53,4 @@ Community resources
     sdks-apis
     ../resources/dashboards
     ../references/developer-tools
+    ../references/grpc
```
Comment (abizjak) [inline source/mainnet/net/guides/developer-page.rst:56]: We need to update the README to describe how the schemas were generated so that the page can be maintained.

File: source/mainnet/net/references/grpc.rst [hunk 6]
```diff
@@ -0,0 +1,886 @@
...
+.. _NodeInfoResponse: https://github.com/Concordium/concordium-grpc-api/blob/44e9c5825b1b18d9e81d15db30546316aa5906ec/concordium_p2p_rpc.proto#L67
+.. |NodeInfoResponse| replace:: ``NodeInfoResponse``
+.. _BlockHeight: _https://github.com/Concordium/concordium-grpc-api/blob/44e9c5825b1b18d9e81d15db30546316aa5906ec/concordium_p2p_rpc.proto#L146
+.. |BlockHeight| replace:: ``BlockHeight``
+.. _PeerElement: https://github.com/Concordium/concordium-grpc-api/blob/44e9c5825b1b18d9e81d15db30546316aa5906ec/concordium_p2p_rpc.proto#L34
+.. |PeerElement| replace:: ``PeerElement``
+.. _PeerStatsResponse: https://github.com/Concordium/concordium-grpc-api/blob/44e9c5825b1b18d9e81d15db30546316aa5906ec/concordium_p2p_rpc.proto#L51
+.. |PeerStatsResponse| replace:: ``PeerStatsResponse``
+.. _PeerListResponse: https://github.com/Concordium/concordium-grpc-api/blob/44e9c5825b1b18d9e81d15db30546316aa5906ec/concordium_p2p_rpc.proto#L46
+.. |PeerListResponse| replace:: ``PeerListResponse``
```
Comment (abizjak) [inline source/mainnet/net/references/grpc.rst:886]: I think these links need to be updated to point to the more recent commit.

File: source/mainnet/net/references/grpc.rst [hunk 7]
```diff
@@ -0,0 +1,886 @@
...
+
+Transaction encoding
+--------------------
+
+This section describes the ``BlockItem`` used in the `SendTransaction
+<#SendTransaction>`_ .
+It also covers the binary serialization of an ``BlockItem`` as that is the
+expected format.
+All possible transactions are *not* covered on this page.
+Instead, it will **focus on transfers and the smart contract-related transactions**.
```
Comment (abizjak) [inline source/mainnet/net/references/grpc.rst:673]: What is "it"?

File: source/mainnet/net/references/grpc.rst [hunk 8]
```diff
@@ -0,0 +1,886 @@
...
+.. function:: PeerVersion() -> String
+
+   Get the version of the node software.
+
+   :returns: The version of the node software.
+   :rtype: String
+
+.. function:: PeerUptime() -> UInt64
+
+   Get the uptime of the *node* in milliseconds.
```
Comment (abizjak) [inline source/mainnet/net/references/grpc.rst:425]: I'm not sure what the reason for putting node in italics is here, and in the 2 calls below.

File: source/mainnet/net/references/grpc.rst [hunk 9]
```diff
@@ -0,0 +1,886 @@
...
+
+   .. collapse:: View JSON schema
+
+      .. literalinclude:: grpc-json-schemas/GetInstanceInfo.json
+         :language: json
+
+.. function:: InvokeContract(blockHash, contractContext) -> ?InvokeContractResult
+
+   Invoke a smart contract instance and view its results as if it had been
+   updated at the end of the given block. *This is not a transaction*.
```
Comment (abizjak) [inline source/mainnet/net/references/grpc.rst:312]: I would recommend elaborating on "This is not a transaction".

File: source/mainnet/net/references/grpc.rst [hunk 10]
```diff
@@ -0,0 +1,886 @@
...
+
+      .. literalinclude:: grpc-json-schemas/GetNextAccountNonce.json
+         :language: json
+
+Smart contracts
+===============
+
+.. function:: GetModuleList(blockHash) -> ?[ModuleHash]
+
+   Get a list of all smart contract modules that existed when the given block was created.
```
Comment (abizjak) [inline source/mainnet/net/references/grpc.rst:251]: The phrasing "when the given block was created" is a bit too imprecise.

I would rephrase as 
```
   Get a list of all smart contract modules that exist in the state at the end of the given block.
```
and the same below.

File: source/mainnet/net/references/grpc.rst (hunk 8 above)
Comment (Bargsteen) [reply source/mainnet/net/references/grpc.rst:425]: I think the name `PeerUptime` is a bit misleading, so I wanted to clarify that it refers to the node itself as opposed to a peer.

File: source/mainnet/net/references/grpc.rst (hunk 9 above)
Comment (Bargsteen) [reply source/mainnet/net/references/grpc.rst:312]: I've added some more text. Let me know if that suffices.

File: source/mainnet/net/references/grpc.rst (hunk 7 above)
Comment (Bargsteen) [reply source/mainnet/net/references/grpc.rst:673]: Rephrased the section.

File: source/mainnet/net/references/grpc.rst (hunk 6 above)
Comment (Bargsteen) [reply source/mainnet/net/references/grpc.rst:886]: Yep, I'll do that once the other PR is merged.

File: source/mainnet/net/references/grpc.rst (hunk 1 above)
Comment (Bargsteen) [reply source/mainnet/net/references/grpc.rst]: What I meant was that the schemas only handled `a` and not `?a`, which wasn't a great solution.
So, as we discussed, I've updated the schemas so they now handle `?a` for the relevant endpoints. (I've queried the node via `concordium-client raw` to determine which endpoints might return null.)

File: source/mainnet/net/references/grpc.rst (hunk 6 above)
Comment (Bargsteen) [reply source/mainnet/net/references/grpc.rst]: Done

**Review by abizjak — APPROVED** [review]

File: source/mainnet/net/references/grpc.rst [hunk 11]
```diff
@@ -0,0 +1,896 @@
...
+
+.. function:: GetInstances(blockHash) -> ?[ContractAddress]
+
+   Get a list of all smart contract instances that exist in the state at the end
+   of the given block.
+
+   :param blockHash: The given block.
+   :type blockHash: |grpc-block-hash|_
+   :returns: A list of smart contract addresses.
+   :rtype: ``[ContractAddress]`` (see JSON schema below)
```
Comment (td202) [inline source/mainnet/net/references/grpc.rst:292]:
```suggestion
   :rtype: ``?[ContractAddress]`` (see JSON schema below)
```

**Review by td202 — COMMENTED** [review]
The format of function declarations is quite sketchy. The names of the arguments are given (converted to camelCase) but not there types; while the type of the return value is given, but as the JSON schema type - except when the return type is not JSON. I think it would be better to more clearly delineate between the protobuf structure and the JSON structure.

In cases where the result can be `null`, it might be useful to describe the circumstances in which `null` is a possible result.

File: source/mainnet/net/references/grpc.rst [hunk 12]
```diff
@@ -0,0 +1,886 @@
...
+These are the SDKs we currently provide:
+
+- `Concordium Rust SDK <https://github.com/Concordium/concordium-rust-sdk>`_
+- `Concordium Javascript (Node / Web) SDK <https://github.com/Concordium/concordium-node-sdk-js>`_
+- `Concordium Java SDK <https://github.com/Concordium/concordium-java-sdk>`_
+- `Concordium .NET (C#) SDK <https://github.com/Concordium/concordium-net-sdk>`_
+- `Concordium Go SDK <https://github.com/Concordium/concordium-go-sdk>`_
+
+The gRPC interface uses a mixture of protobuf defined types and JSON. For the
+JSON types, a `JSON Schema <https://json-schema.org/>`_ is provided.
```
[the remark this answers is not in the document]
Comment (Bargsteen) [reply source/mainnet/net/references/grpc.rst:22]: Added link.

File: source/mainnet/net/references/grpc-json-schemas/GetAccountInfo.json [hunk 13]
```diff
@@ -0,0 +1,816 @@
...
+                "policy": {
+                  "description": "Policy of this credential object.",
+                  "allOf": [
+                    {
+                      "$ref": "#/definitions/Policy_for_HexString_and_AttributeKind"
+                    }
+                  ]
+                },
+                "revocationThreshold": {
+                  "description": "Anonymity revocation threshold. Must be <= length of ar_data.",
```
[the remark this answers is not in the document]
Comment (Bargsteen) [reply source/mainnet/net/references/grpc-json-schemas/GetAccountInfo.json:140]: It is called `ar_data` in the Rust SDK, so I'll leave it as is.

File: source/mainnet/net/references/grpc-json-schemas/ContractContext.json [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    "contract": {
+      "description": "Contract to invoke.",
+      "allOf": [
+        {
+          "$ref": "#/definitions/ContractAddress"
+        }
+      ]
+    },
+    "energy": {
+      "description": "And what amount of energy to allow for execution. This should be small enough so that it can be converted to interpreter energy.",
```
[the remark this answers is not in the document]
Comment (Bargsteen) [reply source/mainnet/net/references/grpc-json-schemas/ContractContext.json:29]: Added a schema restraint to be the maximum energy allowed for invocations.

**Review by abizjak — APPROVED** [review]

File: source/mainnet/net/references/grpc.rst [hunk 15]
```diff
@@ -54,45 +56,47 @@ Transactions
...
    :ref:`grpc-transaction-encoding` section.
 
-   :param Int32 networkId: The network that the transaction should be sent to
+   :param Int32 network_id: The network that the transaction should be sent to
                            (only ``100`` is currently supported).
    :param payload: Binary encoding of the transaction payload.
-   :type payload: |grpc-block-item|_
+   :type payload: ``[Byte]`` of |grpc-block-item|_
    :returns: Whether the transaction succeeded.
    :rtype: Bool
```
Comment (td202) [inline source/mainnet/net/references/grpc.rst:72]: "succeeded" is a somewhat ambiguous term. In fact, this returns `True` if the transaction was successfully submitted, and never returns `False`. If there was a problem with the transaction, a GRPC error code will be used:

- The `SendTransaction` function exposed via the gRPC interface now provides the caller with detailed error messages if the 
  transaction was rejected instead of just `False`. The function still returns `True` if 
  the transaction was accepted.
  The following gRPC error codes can be returned.
  - 'SUCCESS' The transaction was succesfully relayed to consensus.
  - 'INVALID_ARGUMENT' The transaction was deemed invalid or exceeds the maximum size allowed (the raw size of the transaction).
     In addition the error message contains information as to why the transaction was deemed invalid.
  - 'FAILED_PRECONDITION' The network was stopped due to an unrecognized protocol update.
  - 'DUPLICATE_ENTRY' The transaction was a duplicate.
  - 'INTERNAL' An internal error happened and as such the transaction could not be processed.
  The server will return a gRPC status if the transaction was deemed invalid.

File: source/mainnet/net/references/grpc.rst [hunk 16]
```diff
@@ -514,21 +535,21 @@ Networks and peers
    :returns: Whether the request was processed successfully.
    :rtype: Bool
 
-.. function:: BanNode(peerElement) -> Bool
+.. function:: BanNode(peer_element: PeerElement) -> Bool
```
Comment (td202) [inline source/mainnet/net/references/grpc.rst:538]: This seems inconsistent with other cases, where the message type (PeerElement) is expanded into its fields. (Also UnbanNode.)

File: source/mainnet/net/references/grpc.rst [hunk 17]
```diff
@@ -468,33 +488,34 @@ The node
...
-.. function:: PeerList(includeBootstrappers) -> PeerListResponse
+.. function:: PeerList(include_bootstrappers: Bool) -> PeerListResponse
 
    Get a list of the peers that the node is connected to.
 
-   :param Bool includeBootstrappers: Whether to include the bootstrapper nodes
+   :param Bool include_bootstrappers: Whether to include the bootstrapper nodes
                                      in the response.
    :returns: A list of peers.
    :rtype: |PeerListResponse|_
```
Comment (td202) [inline source/mainnet/net/references/grpc.rst:507]: Perhaps give more details of the fields of the result message?

**Review by td202 — APPROVED** [review]

File: source/mainnet/net/references/grpc.rst (hunk 17 above)
Comment (Bargsteen) [reply source/mainnet/net/references/grpc.rst:507]: The fields are described in the protobuf file that is linked.  Is that not sufficient?

File: source/mainnet/net/references/grpc.rst [hunk 18]
```diff
@@ -68,7 +68,16 @@ Transactions
...
-   :returns: Whether the transaction succeeded.
+   :returns: Either ``True`` or one of the following gRPC errors:
+
+               - ``INVALID_ARGUMENT``: The transaction was deemed invalid or
+                 exceeds the maximum size allowed (the raw size of the transaction).
+               - ``FAILED_PRECONDITION``: The network was stopped due to an
+                 unrecognized protocol update.
+               - ``DUPLICATE_ENTRY``: The transaction was a duplicate.
+               - ``INTERNAL``: An internal error happened and as such the
+                 transaction could not be processed. The node will retrun a gRPC
```
Comment (td202) [inline source/mainnet/net/references/grpc.rst:79]:
```suggestion
                 transaction could not be processed. The node will return a gRPC
```

File: source/mainnet/net/references/grpc.rst (hunk 17 above)
Comment (td202) [reply source/mainnet/net/references/grpc.rst:516]: I suppose so.

**Review by td202 — APPROVED** [review]
